File: typescript/tests/integration/session-idle-release-timeout.integration.test.ts
import { AgentBay } from "../../src";
import { getTestApiKey } from "../utils/test-helpers";
import { log } from "../../src/utils/logger";

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe("Session idle release timeout (integration)", () => {
  it("should release session between timeout and timeout+60s", async () => {
    const apiKey = getTestApiKey();
    const agentBay = new AgentBay({ apiKey });

    const idleReleaseTimeoutSeconds = 60;
    const maxOverSeconds = 60;
    const pollIntervalMs = 2000;
    const imageId = "computer-use-ubuntu-2204-regionGW";

    const createStart = Date.now();
    const createResult = await agentBay.create({
      imageId,
      idleReleaseTimeout: idleReleaseTimeoutSeconds,
      labels: {
        test: "idle-release-timeout",
        sdk: "typescript",
      },
    });

    expect(createResult.success).toBe(true);
    expect(createResult.session).toBeDefined();
    const session = createResult.session!;

    try {
      const start = Date.now();
      const timeoutDeadline = start + idleReleaseTimeoutSeconds * 1000;

      while (Date.now() < timeoutDeadline) {
        const status = await session.getStatus();
        if (!status.success && status.code === "InvalidMcpSession.NotFound") {
          throw new Error(`Session released too early: NotFound before ${idleReleaseTimeoutSeconds}s`);
        }
        if (status.success && ["FINISH", "DELETING", "DELETED"].includes(status.status)) {
          throw new Error(`Session released too early: status=${status.status} before ${idleReleaseTimeoutSeconds}s`);
        }
        const remaining = timeoutDeadline - Date.now();
        await sleep(Math.min(pollIntervalMs, Math.max(0, remaining)));
      }

      const deadline = timeoutDeadline + maxOverSeconds * 1000;
      while (Date.now() < deadline) {
        const status = await session.getStatus();

        if (!status.success && status.code === "InvalidMcpSession.NotFound") {
          const elapsed = (Date.now() - start) / 1000;
          expect(elapsed).toBeGreaterThanOrEqual(idleReleaseTimeoutSeconds);
          expect(elapsed).toBeLessThanOrEqual(idleReleaseTimeoutSeconds + maxOverSeconds);
          log(`✅ Session released: NotFound, elapsed=${elapsed.toFixed(2)}s`);
          return;
        }

        if (status.success && ["FINISH", "DELETING", "DELETED"].includes(status.status)) {
          const elapsed = (Date.now() - start) / 1000;
          expect(elapsed).toBeGreaterThanOrEqual(idleReleaseTimeoutSeconds);
          expect(elapsed).toBeLessThanOrEqual(idleReleaseTimeoutSeconds + maxOverSeconds);
          log(`✅ Session released: status=${status.status}, elapsed=${elapsed.toFixed(2)}s`);
          return;
        }

        await sleep(pollIntervalMs);
      }

      throw new Error(
        `Session was not released within ${idleReleaseTimeoutSeconds}s~${idleReleaseTimeoutSeconds + maxOverSeconds}s`
      );
    } finally {
      // Best-effort cleanup (session may already be released)
      try {
        const status = await session.getStatus();
        if (status.success && !["FINISH", "DELETING", "DELETED"].includes(status.status)) {
          await session.delete();
        }
      } catch (e) {
        // ignore
      }
      log(`Create request started at: ${new Date(createStart).toISOString()}`);
    }
  }, 2 * 60 * 1000);
});

File: golang/tests/pkg/integration/session_idle_release_timeout_integration_test.go
package integration

import (
	"fmt"
	"os"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
)

func isNotFoundStatusResult(r *agentbay.SessionStatusResult) bool {
	if r == nil {
		return false
	}
	if r.Success {
		return false
	}
	return r.Code == "InvalidMcpSession.NotFound"
}

// TestSessionIdleReleaseTimeoutIntegration verifies session is released between timeout and timeout+60s.
func TestSessionIdleReleaseTimeoutIntegration(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	idleReleaseTimeoutSeconds := int32(60)
	maxOverSeconds := int32(60)
	imageID := "linux_latest"

	params := agentbay.NewCreateSessionParams().
		WithImageId(imageID).
		WithIdleReleaseTimeout(idleReleaseTimeoutSeconds).
		WithLabels(map[string]string{
			"test": "idle-release-timeout",
			"sdk":  "golang",
		})

	start := time.Now()
	createResult, err := client.Create(params)
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	if createResult == nil || createResult.Session == nil {
		t.Fatalf("Create returned nil session")
	}
	session := createResult.Session
	t.Logf("Session created: %s", session.SessionID)

	defer func() {
		// Best-effort cleanup: if still running, delete it.
		status, e := session.GetStatus()
		if e == nil && status != nil && status.Success && status.Status != "" &&
			status.Status != "FINISH" && status.Status != "DELETING" && status.Status != "DELETED" {
			_, _ = session.Delete()
		}
	}()

	timeoutDeadline := start.Add(time.Duration(idleReleaseTimeoutSeconds) * time.Second)
	for time.Now().Before(timeoutDeadline) {
		status, e := session.GetStatus()
		if e != nil {
			t.Fatalf("GetStatus failed: %v", e)
		}
		if isNotFoundStatusResult(status) {
			t.Fatalf("Session released too early: NotFound before %ds", idleReleaseTimeoutSeconds)
		}
		if status.Success && (status.Status == "FINISH" || status.Status == "DELETING" || status.Status == "DELETED") {
			t.Fatalf("Session released too early: status=%s before %ds", status.Status, idleReleaseTimeoutSeconds)
		}
		remaining := time.Until(timeoutDeadline)
		sleep := 2 * time.Second
		if remaining < sleep {
			sleep = remaining
		}
		time.Sleep(sleep)
	}

	deadline := timeoutDeadline.Add(time.Duration(maxOverSeconds) * time.Second)
	for time.Now().Before(deadline) {
		status, e := session.GetStatus()
		if e != nil {
			t.Fatalf("GetStatus failed: %v", e)
		}

		if isNotFoundStatusResult(status) {
			elapsed := time.Since(start).Seconds()
			if elapsed < float64(idleReleaseTimeoutSeconds) {
				t.Fatalf("Session released too early: elapsed=%.2fs", elapsed)
			}
			if elapsed > float64(idleReleaseTimeoutSeconds+maxOverSeconds) {
				t.Fatalf("Session released too late: elapsed=%.2fs", elapsed)
			}
			fmt.Printf("✅ Session released: NotFound, elapsed=%.2fs\n", elapsed)
			return
		}

		if status.Success && (status.Status == "FINISH" || status.Status == "DELETING" || status.Status == "DELETED") {
			elapsed := time.Since(start).Seconds()
			if elapsed < float64(idleReleaseTimeoutSeconds) {
				t.Fatalf("Session released too early: elapsed=%.2fs", elapsed)
			}
			if elapsed > float64(idleReleaseTimeoutSeconds+maxOverSeconds) {
				t.Fatalf("Session released too late: elapsed=%.2fs", elapsed)
			}
			fmt.Printf("✅ Session released: status=%s, elapsed=%.2fs\n", status.Status, elapsed)
			return
		}

		time.Sleep(2 * time.Second)
	}

	t.Fatalf("Session was not released within %ds~%ds", idleReleaseTimeoutSeconds, idleReleaseTimeoutSeconds+maxOverSeconds)
}
